src/python/pants/build_graph/build_configuration.py

```diff
@@ -34,7 +34,7 @@ def __init__(self):
     self._exposed_context_aware_object_factory_by_alias = {}
     self._optionables = OrderedSet()
     self._rules = OrderedSet()
-    self._union_rules = OrderedDict()
+    self._union_rules = {}
```

Comment:
There is no use for ordered dictionaries in Python 3.6+ due to normal dictionaries being ordered by insertion order. Normal dicts are more ergonomic than `OrderedDict`, e.g. better `repr()` and simpler type hints.

Reply:
`The order-preserving aspect of this new implementation is considered an implementation detail and should not be relied upon`

https://docs.python.org/3.6/whatsnew/3.6.html#other-language-changes

Reply:
Yes, but they formalized it in Python 3.7 and we only run Pants with CPython so we are certain this is how Python 3.6 will behave.

Iirc, they put that disclosure because they weren't sure if they wanted to commit to dictionaries always being ordered for the rest of Python's history. This ordering started as a side effect of a performance improvement. Then, in 3.7, they confirmed yes this should always be the case.

Reply:
This nonetheless seems like it removes documentation, and is unrelated to the change. Not a blocker, but IMO: leaving OrderedDict in place where we are "definitely relying on ordering" would be a good thing to do in the future.